### PR TITLE
[chonky] Fix #6: restructure frontend content

### LINKED ISSUE #6: Restructure content?
I'm pretty sure that to make content legible, each "page" / view should contain docs on exactly one top-level... thing. So for example
* A module page may contain a listing of all the top-level symbols, maybe with the first line or so of its docstring
  * Maybe variable / constant docs are just fully inlined here
* Separate pages for classes (what's up with inner classes?)
* Separate pages for top-level functions?

I don't know what plans you had in mind about all this. I've hinted at this before, but I'll just drop it here explicitly so that there's a place to discuss it: my default approach would be to model the organization structure on docs.rs. Some things obviously need adapting as the languages differ, but even so that should be much less effort than trying to design this from first principles.

### PR DESCRIPTION
**Pull request by abesto**
This is a big huge commit. Somewhat sorry. Halfway through I realized I needed some extra handholding from linters / formatters, so that's just ... also inline here. Quite sorry.

One significant departure from what was discussed on #6: I added some more structure to the URLs (see below). This makes the logic of finding the right symbol way simpler; you don't need to do stuff like "find the longest prefix of this symbol that's a module in this package". This means that given nested modules and inner classes, we may have a URL like `mypkg/moda.modb/ClassA.ClassB`. This is *much* easier to parse than `mypkg.moda.modb.ClassA.ClassB`.

`/` didn't change, except it now shows the version number.

<details>
  <summary><code>/click</code> (<code>/[pkg]</code>) shows a mostly-empty sidebar, and the list of modules in the package.</summary

image

</details>

Clicking the "Package $PKG" in the top left goes back to the package overview.

<details>
  <summary><code>/click/decorators</code> (<code>/[pkg]/[mod]</code>) shows a list of modules in the sidebar, and lists the contents of the module in the main pane.</summary>

*  The current module is highlighted in the sidebar.
* If there are duplicate entries, only one item is shown for each name; if there's one with non-empty documentation, then it's that. (If there are multiple with non-empty documentation, then: the first one, sorted by description. Or the last one. Ordering is confusing.)

image

</details>

<details>
  <summary><code>/click/decorators/group</code> (<code>/[pkg]/[mod]/[symbol]</code> for a function) shows module contents in the sidebar, and the function (including all its incarnations) in the main pane.</summary>

(Variables behave the same way)

image

</details>

<details>
  <summary><code>/click/core/Command</code> (<code>/[pkg]/[mod]/[symbol]</code> for a class) lists the class contents, _then_ the module contents in the sidebar, and the fully expanded class contents in the main pane.</summary>

Links to class members don't work yet (that's for #3 )

image

</details>

### CODE REVIEW

**Comment by abesto** [comment]
Hah. Bot that commits a fix >>> bot that complains

File: www/.eslintrc.json [hunk 1]
```diff
@@ -1,3 +1,28 @@
 {
-  "extends": "next/core-web-vitals"
+  "parser": "@typescript-eslint/parser",
```
Comment (zsol) [inline www/.eslintrc.json:2]: gotta figure out how to run this in CI

File: www/lib/utils.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+export function cmpBy<T>(...key: (keyof T)[]) {
```
Comment (zsol) [inline www/lib/utils.ts:1]: can we `utils.ts` -> `sorting.ts`? I have an aversion to generic names like `utils` and `common` and `helpers`

File: www/lib/url.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import { UrlObject } from "url";
+
+import * as docs from "./docs";
+
+export function withoutPrefix(prefix: string, s: string) {
+  if (s.startsWith(prefix)) {
```
Comment (zsol) [inline www/lib/url.ts:6]:
```suggestion
  if (s.startsWith(`${prefix}.`)) {
```

File: www/components/Docs/SymbolLinkTable.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+function deduplicate<T extends Sym>(xs: T[]): T[] {
+  return sortedBy(xs, "name", "documentation")
+    .reverse()
+    .filter((x, i, xs) => {
+      return i === 0 || x.name !== xs[i - 1].name;
+    })
+    .reverse();
+}
```
Comment (zsol) [inline www/components/Docs/SymbolLinkTable.tsx:32]: I feel ambivalent about this. Apart from `@overload` this shouldn't be necessary especially after #19 
For `@overload`s, there will need to be some special handling: we don't want to drop overloads as they meaningfully help understand the API (probably more than the signature of the implementation).

File: www/components/Docs/SymbolLinkTable.tsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    <Box>
+      <Text variant="h3">{title}</Text>
+      <Table>
+        <Table.Tbody>
+          {deduplicate(symbols).map((sym) => {
+            return (
+              <Table.Tr key={sym.name}>
+                <Table.Th>
+                  <NextLink href={url(sym)} passHref>
+                    <Link>{sym.name}</Link>
```
Comment (zsol) [inline www/components/Docs/SymbolLinkTable.tsx:52]: We should strip the module prefix from this name - they don't provide much value but increase the size of the table

**Review by zsol — APPROVED** [review]

**Comment by zsol** [comment]
How will nested classes work with this new url scheme?

File: www/lib/utils.ts (hunk 2 above)
Comment (abesto) [reply www/lib/utils.ts:1]: Samesies. I had various pieces of code in here at various points of the refactor. `sorting` is good, will do.

File: www/components/Docs/SymbolLinkTable.tsx (hunk 4 above)
Comment (abesto) [reply www/components/Docs/SymbolLinkTable.tsx:32]: I implemented this specifically for functions. Agree that once (if) we figure out a way to uniquely represent each signature, this should go away. That needs some care - for example, `click` only adds docstrings on one of the versions.

**Comment by abesto** [comment]
> How will nested classes work with this new url scheme?

They should Just Work, but I'd love a test-case (I mean: a library that has them) if you have one handy. At least in theory, with nested modules and inner classes, we may have a URL like `mypkg/moda.modb/ClassA.ClassB`. (And then members of the class would be either `mypkg/moda.modb.ClassA.ClassB#method_a` or `mypkg/moda.modb.ClassA.ClassB?hl=method_a`, depending on what I can convince NextJS to do)

File: www/.eslintrc.json (hunk 1 above)
Comment (abesto) [reply www/.eslintrc.json:2]: Tru dat. Filed https://github.com/zsol/py.wtf/issues/21

**Comment by abesto** [comment]
image

**Comment by abesto** [comment]
Actually, never mind all that, `pre-commit.ci` is not a good fit for our ESLint because we have some rules in there that require all NPM dependencies to be installed, so that's more a job for #23

**Review by zsol — APPROVED** [review]

**Pull request merged**